contracts/near/src/lib.rs

```diff
@@ -0,0 +1,489 @@
+///
+/// There is no valid Default state for this contract, so we derive PanicOnDefault to force
+/// deployment using one of the #[init] functions in the impl below.
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+pub struct Pyth {
+    /// The set of `Source`s from which Pyth attestations are allowed to be relayed from.
+    pub sources: UnorderedSet<Source>,
+
+    /// The Governance Source.
+    pub gov_source: Source,
+
+    /// The last executed sequence number for governance actions.
+    pub last_gov_seq: u64,
+
+    /// A Mapping from PriceFeed ID to Price Info.
+    pub prices: UnorderedMap<Identifier, PriceFeed>,
+
+    /// The `AccountId` of the Wormhole contract used to verify VAA's.
+    pub wormhole: AccountId,
+
+    /// A hash of the current contract code.
+    ///
+    /// If this `hash` does not match the current contract code, then it indicates that a pending
+    /// upgrade has been allowed, calling `upgrade` with code that matches this hash will cause the
+    /// contract to upgrade itself.
+    pub codehash: [u8; 32],
+
+    /// Staleness threshold for rejecting price updates.
+    pub stale_threshold: u64,
+
+    /// Fee for updating price.
+    pub update_fee: Balance,
+}
+
+/// The `Action` acts as an enum of possible `Action`'s this contract can receive via VAA.
+#[derive(BorshDeserialize, BorshSerialize)]
+pub enum Action {
+    BatchAttest(Vec<u8>),
+    ContractUpgrade([u8; 32]),
+    SetDataSources(Vec<Source>),
+    SetGovernanceSource(Source),
+    SetStalePriceThreshold(u64),
+    SetUpdateFee(Balance),
+}
+
+#[near_bindgen]
+impl Pyth {
+    #[init]
+    #[allow(clippy::new_without_default)]
+    pub fn new(
+        wormhole: AccountId,
+        codehash: [u8; 32],
+        initial_source: Source,
+        gov_source: Source,
+        update_fee: Balance,
+        stale_threshold: u64,
+    ) -> Self {
+        // Add an initial Source so that the contract can be used.
+        let mut sources = UnorderedSet::new(StorageKeys::Source);
+        sources.insert(&initial_source);
+        Self {
+            prices: UnorderedMap::new(StorageKeys::Prices),
+            last_gov_seq: 0,
+            stale_threshold,
+            gov_source,
+            sources,
+            wormhole,
+            codehash,
+            update_fee,
+        }
+    }
+
+    #[init(ignore_state)]
+    pub fn migrate() -> Self {
+        let state: Self = env::state_read().expect("Failed to read state");
+        state
+    }
+
+    /// Instruction for processing VAA's relayed via Wormhole.
+    ///
+    /// Note that VAA verification requires calling Wormhole so processing of the VAA itself is
+    /// done in a callback handler, see `process_vaa_callback`.
+    #[payable]
+    #[handle_result]
+    pub fn update_price_feeds(&mut self, vaas: Vec<String>) -> Result<(), Error> {
+        for vaa_hex in vaas {
```

Comment:
Can you explain the behaviour when one of the VAAs fail here? Only the successful ones will succeed, right?
Also since we are in a loop, are gas conditions accurate?

Reply:
You were right about this, and unfortunately It's not possible to wait on a promise chain in a reasonable way to get the deposit refunds correct, I tried several ways of combining the promises but without threading context through everything into a final handler it just gets lost. I decided It's not worth the complexity risk and callers, should they need it, can simply chain call the update function if they need several. The costs should be similar.